util/EIP-2124/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! EIP-2124 implementation based on <https://eips.ethereum.org/EIPS/eip-2124>
+
+use crc::crc32;
+use ethereum_types::H256;
+use maplit::*;
+use rlp::{DecoderError, Rlp, RlpStream};
+use rlp_derive::*;
+use std::collections::{BTreeMap, BTreeSet};
+
+pub type BlockNumber = u64;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct ForkHash(pub u32);
+
+impl rlp::Encodable for ForkHash {
+	fn rlp_append(&self, s: &mut RlpStream) {
+		s.encoder().encode_value(&self.0.to_be_bytes());
+	}
+}
+
+impl rlp::Decodable for ForkHash {
+	fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+		rlp.decoder().decode_value(|b| {
+			if b.len() != 4 {
+				return Err(DecoderError::RlpInvalidLength);
+			}
+
+			let mut blob = [0; 4];
+			blob.copy_from_slice(&b[..]);
+
+			return Ok(Self(u32::from_be_bytes(blob)))
+		})
+	}
+}
+
+impl From<H256> for ForkHash {
+	fn from(genesis: H256) -> Self {
+		Self(crc32::checksum_ieee(&genesis[..]))
+	}
+}
+
+impl std::ops::AddAssign<BlockNumber> for ForkHash {
+	fn add_assign(&mut self, height: BlockNumber) {
+		let blob = height.to_be_bytes();
+		self.0 = crc32::update(self.0, &crc32::IEEE_TABLE, &blob)
+	}
+}
+
+impl std::ops::Add<BlockNumber> for ForkHash {
+	type Output = Self;
+	fn add(mut self, height: BlockNumber) -> Self {
+		self += height;
+		self
+	}
+}
+
+/// A fork identifier as defined by EIP-2124.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, RlpEncodable, RlpDecodable)]
+pub struct ForkId {
+	/// CRC32 checksum of the all fork blocks from genesis.
+	pub hash: ForkHash,     
+	/// Next upcoming fork block number, 0 if not yet known.
+	pub next: BlockNumber
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum RejectReason {
+	RemoteStale,
+	LocalIncompatibleOrStale,
+}
+
+#[derive(Clone, Debug)]
+pub struct ForkFilter {
+	pub head: BlockNumber,
+	past_forks: BTreeMap<BlockNumber, ForkHash>,
+	next_forks: BTreeSet<BlockNumber>,
+}
+
+impl ForkFilter {
+	pub fn new<PF, NF>(head: BlockNumber, genesis: H256, past_forks: PF, next_forks: NF) -> Self
+	where
+		PF: IntoIterator<Item = BlockNumber>,
+		NF: IntoIterator<Item = BlockNumber>,
+	{
+		let genesis_fork_hash = ForkHash::from(genesis);
+		Self {
+			head,
+			past_forks: past_forks.into_iter().fold((btreemap! { 0 => genesis_fork_hash }, genesis_fork_hash), |(mut acc, base_hash), block| {
+				let fork_hash = base_hash + block;
+				acc.insert(block, fork_hash);
+				(acc, fork_hash)
+			}).0,
+			next_forks: next_forks.into_iter().collect(),
+		}
+	}
+
+	fn current_fork_hash(&self) -> ForkHash {
+		*self.past_forks.values().next_back().unwrap()
```

Comment:
No `unwrap()`s outside test code please. If we are sure the code cannot fail it's ok to use `.expect("this never fails because: x and y and z; qed");`.

Reply:
done